### PR TITLE
#66: [Persist emails before send] persist messages before send (closes #66)

### LINKED ISSUE #66: [Persist emails before send] persist messages before send
← #59

## requirements
persist messages before send

## specs
{optional: describe technical specs to implement this feature, if not obvious}

## misc
{optional: other useful info}

### PR DESCRIPTION
**Pull request by calippo**
Closes #66

### tests performed

> A Test Plan is used to show what you tested to make sure your code works fine. You should write here all that you did to test, and provide some results of your testing.

> These results can be screenshots, query results, or even just pointers to unit tests that successfully passed.

### tests not performed (domain coverage)

> At times not everything can be tested, and writing what hasn't been tested is just as important as writing what has been tested.

> An example of partial test is a field displaying 4 possible values. If 3 values are tested, with screenshots, and 1 is not, then it should be mentioned here.

### CODE REVIEW

File: src/main/scala/mailo/persistence/EmailPersistanceActor.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    case SnapshotOffer(_, snapshot: EmailState) => state = snapshot
+  }
+
+  private[this] val snapShotInterval = 1000
+
+  val receiveCommand: Receive = {
+    case command@CleanQueue(sequenceNr) =>
+      log.info("received event {}", command.toString)
+      saveSnapshot(state)
+      deleteMessages(sequenceNr)
```
Comment (calippo) [inline src/main/scala/mailo/persistence/EmailPersistanceActor.scala]: honestly I'm not sure about the clean queue part, need to better test it

File: src/main/scala/mailo/persistence/CustomContentTypeCodecs.scala [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  implicit val encodeFoo: Encoder[ContentType] = new Encoder[ContentType] {
+    final def apply(c: ContentType): Json =
+      Json.obj(("contentType", Json.fromString(c.value)))
+  }
+
+  implicit val decodeFoo: Decoder[ContentType] = new Decoder[ContentType] {
+    final def apply(c: HCursor): Decoder.Result[ContentType] =
+      for {
+        contentType <- c.downField("contentType").as[String]
+        result <- ContentType.parse(contentType).left.map(_ => DecodingFailure("Falied decoding content type", Nil))
```
Comment (bytecodeguru) [inline src/main/scala/mailo/persistence/CustomContentTypeCodecs.scala]: Typo `Falied` -> `Failed`

File: src/main/scala/mailo/persistence/SnapshotHelper.scala [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package mailo.persistence
+
+import akka.persistence._
+
+import com.typesafe.scalalogging.LazyLogging
+
+object SnapshotHelper extends LazyLogging {
+  val logSnapshotResult: PartialFunction[Any, Unit] = {
+    case SaveSnapshotSuccess(metadata) =>
+      logger.debug(s"Snaposhot successfully stored")
```
Comment (bytecodeguru) [inline src/main/scala/mailo/persistence/SnapshotHelper.scala]: typo `Snaposhot` -> `Snapshot`

**Review by bytecodeguru — COMMENTED** [review]
I need an offline explanation of the persistence machinery.

I've found just a couple of typos for now.

**Comment by calippo** [comment]
@bytecodeguru as discussed offline, updated macro issue with the following:
- convert dry run to integration test
- add dead letter queue
- add queue monitoring api

https://github.com/buildo/mailo/issues/59

File: src/main/resources/reference.conf [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+#     serializers {
+#       java = "akka.serialization.JavaSerializer"
+#       proto = "akka.remote.serialization.ProtobufSerializer"
+#     }
+
+#     serialization-bindings {
+#       "mailo.persistence.SendEmail" = proto
+#     }
+#   }
+# }
```
Comment (bytecodeguru) [inline src/main/resources/reference.conf:18]: Why is this commented out?

File: src/main/scala/mailo/AtLeastOnceMailo.scala [hunk 5]
```diff
@@ -1,40 +1,39 @@
...
+  private[this] val emailPersistanceActor = system.actorOf(EmailPersistanceActor.props(emailSender, loggingActor))
 
-  def send(mail: Mail) = {
-    ask.map {
-      case Success(result) => Right(LocallyQueued)
-      case Failure(error)  => Left(MailPersistenceError(error.getLocalizedMessage))
-    }
+  def send(mail: Mail): Future[Either[MailError, MailResult]] = {
+    ask(emailPersistanceActor, SendEmail(mail))
+      .map(_ => Right(LocallyQueued))
```
Comment (bytecodeguru) [inline src/main/scala/mailo/AtLeastOnceMailo.scala]: Is it really safe to ignore `_`?

File: src/main/scala/mailo/DryRun.scala [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    subject: String,
+    content: MailRefinedContent.MailRefinedContent,
+    attachments: List[Attachment],
+    tags: List[String],
+    headers: Map[String, String]
+  )(implicit executionContext: ExecutionContext): Future[Either[MailError, MailResponse]] =
+    Future(Right(MailResponse("id", "ok")))
+}
+
+object AtLeastOnceDryRun extends App {
```
Comment (bytecodeguru) [inline src/main/scala/mailo/DryRun.scala:34]: Is this a test?

File: src/main/scala/mailo/persistence/CustomContentTypeCodecs.scala [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package mailo.persistence
+
+import io.circe.{Decoder, Encoder, HCursor, Json}
+import akka.http.scaladsl.model.ContentType
+import io.circe.DecodingFailure
+
+trait CustomContentTypeCodecs {
+  implicit val encodeFoo: Encoder[ContentType] = new Encoder[ContentType] {
```
Comment (bytecodeguru) [inline src/main/scala/mailo/persistence/CustomContentTypeCodecs.scala]: What about `contentTypeEncoder` and `contentTypeDecoder`?

File: src/main/scala/mailo/persistence/EmailPersistanceActor.scala [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+  val logSnapshotResult: Receive = SnapshotHelper.logSnapshotResult
+}
+
+class LoggingActor() extends Actor with ActorLogging with CustomContentTypeCodecs {
+  def receive = {
+    case EmailApplicativeErrorEvent(event, errorMessage) =>
+      log.error(s"%{event} failed with error: ${errorMessage}")
+    case EmailCommunicationErrorEvent(sendEmail, errorMessage) =>
+      log.error(s"${sendEmail} failed with error ${errorMessage}, rescheduling the message")
```
Comment (bytecodeguru) [inline src/main/scala/mailo/persistence/EmailPersistanceActor.scala]: We are not rescheduling anything yet... the logged message is misleading.

**Review by bytecodeguru — CHANGES_REQUESTED** [review]
Nice!
I've left some minor suggestions.

File: src/main/resources/reference.conf (hunk 4 above)
Comment (calippo) [reply src/main/resources/reference.conf:18]: It took me 1 hour to find the proper setting, and currently it's buggish xD

File: src/main/scala/mailo/DryRun.scala (hunk 6 above)
Comment (calippo) [reply src/main/scala/mailo/DryRun.scala:34]: It's a dry run, not a proper test, this is removed in the next pr.

**Comment by calippo** [comment]
@bytecodeguru addressed most of the comments.

this is the only major change
image

Q: "is it safe to discard the result"?
A: "we're now returning the result itself, there shouldn't be any problem (even though I'm not convinces `.mapTo` is super typesafe).

**Review by bytecodeguru — APPROVED** [review]
LGTM 👍

**Pull request merged**